### PR TITLE
Lilly C16 Pine

### PR DESCRIPTION
**Pull request by waterlilly169**

### CODE REVIEW

File: linked_list/linked_list.py [hunk 1]
```diff
@@ -13,63 +13,140 @@ def __init__(self):
 
     # returns the value in the first node
     # returns None if the list is empty
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(1)?
+    # Space Complexity: O(1)?
     def get_first(self):
```
Comment (spitsfire) [inline linked_list/linked_list.py:18]: 👍

File: linked_list/linked_list.py [hunk 2]
```diff
@@ -13,63 +13,140 @@ def __init__(self):
...
+        if self.head == None:
+            return None
 
+        return self.head.value
 
     # method to add a new node with the specific data value in the linked list
     # insert the new node at the beginning of the linked list
     # Time Complexity: ?
     # Space Complexity: ?
     def add_first(self, value):
```
Comment (spitsfire) [inline linked_list/linked_list.py:28]: 👍

File: linked_list/linked_list.py [hunk 3]
```diff
@@ -13,63 +13,140 @@ def __init__(self):
...
 
     # method to find if the linked list contains a node with specified value
     # returns true if found, false otherwise
     # Time Complexity: ?
     # Space Complexity: ?
     def search(self, value):
-        pass
+        current = self.head 
+
+        while current != None:
```
Comment (spitsfire) [inline linked_list/linked_list.py:38]: We can shorten this to:

```suggestion
        while current:
```

File: linked_list/linked_list.py [hunk 4]
```diff
@@ -13,63 +13,140 @@ def __init__(self):
...
 
     # method that returns the length of the singly linked list
     # Time Complexity: ?
     # Space Complexity: ?
     def length(self):
-        pass
+        current = self.head
+
+        count = 0 
+        while current != None:
```
Comment (spitsfire) [inline linked_list/linked_list.py:51]:
```suggestion
        while current:
```

File: linked_list/linked_list.py [hunk 5]
```diff
@@ -13,63 +13,140 @@ def __init__(self):
...
+            if current.value == value:
+                return True 
+            current = current.next
+        return False 
 
     # method that returns the length of the singly linked list
     # Time Complexity: ?
     # Space Complexity: ?
     def length(self):
```
Comment (spitsfire) [inline linked_list/linked_list.py:47]: 👍

File: linked_list/linked_list.py [hunk 6]
```diff
@@ -13,63 +13,140 @@ def __init__(self):
...
+
+        while current != None:
+            if current.value == value:
+                return True 
+            current = current.next
+        return False 
 
     # method that returns the length of the singly linked list
     # Time Complexity: ?
     # Space Complexity: ?
```
Comment (spitsfire) [inline linked_list/linked_list.py]: What do you think the space and time complexity of this is? Is the time complexity dependent on the length of the linked list? Is there any new variables being created that grow larger depending upon the input?

File: linked_list/linked_list.py [hunk 7]
```diff
@@ -13,63 +13,140 @@ def __init__(self):
...
+            count += 1
+        
+        return count
 
     # method that returns the value at a given index in the linked list
     # index count starts at 0
     # returns None if there are fewer nodes in the linked list than the index value
     # Time Complexity: ?
     # Space Complexity: ?
     def get_at_index(self, index):
```
Comment (spitsfire) [inline linked_list/linked_list.py:62]: 👍

File: linked_list/linked_list.py [hunk 8]
```diff
@@ -13,63 +13,140 @@ def __init__(self):
...
+            current = current.next
+            count += 1
+        
+        return count
 
     # method that returns the value at a given index in the linked list
     # index count starts at 0
     # returns None if there are fewer nodes in the linked list than the index value
     # Time Complexity: ?
     # Space Complexity: ?
```
Comment (spitsfire) [inline linked_list/linked_list.py]: What do you think the space and time complexity of this is? Is the time complexity dependent on the length of the linked list? Is there any new variables being created that grow larger depending upon the input?

File: linked_list/linked_list.py [hunk 9]
```diff
@@ -13,63 +13,140 @@ def __init__(self):
...
+    # returns -1 if the value is not found  in the linked list
+    # Time Complexity: O(n)?
+    # Space Complexity: O(1)?
+    
 
     # method that returns the value of the last node in the linked list
     # returns None if the linked list is empty
     # Time Complexity: ?
     # Space Complexity: ?
     def get_last(self):
```
Comment (spitsfire) [inline linked_list/linked_list.py:84]: 👍

File: linked_list/linked_list.py [hunk 10]
```diff
@@ -13,63 +13,140 @@ def __init__(self):
...
+        current = self.head 
+
+        while current.next != None:
+            current = current.next
+
+        return current.value
 
     # method that inserts a given value as a new last node in the linked list
     # Time Complexity: ?
     # Space Complexity: ?
```
Comment (spitsfire) [inline linked_list/linked_list.py]: What do you think the space and time complexity of this is? Is the time complexity dependent on the length of the linked list? Is there any new variables being created that grow larger depending upon the input?

File: linked_list/linked_list.py [hunk 11]
```diff
@@ -13,63 +13,140 @@ def __init__(self):
...
+    #method to add node to the end of the linked list
     def add_last(self, value):
```
Comment (spitsfire) [inline linked_list/linked_list.py:100]: 👍

File: linked_list/linked_list.py [hunk 12]
```diff
@@ -13,63 +13,140 @@ def __init__(self):
...
-        pass
+        current = self.head
+        if current is None:
+            self.head = Node(value)
+            return
+        while current.next != None:
```
Comment (spitsfire) [inline linked_list/linked_list.py:105]:
```suggestion
        while current.next:
```

File: linked_list/linked_list.py [hunk 13]
```diff
@@ -13,63 +13,140 @@ def __init__(self):
...
+            current = current.next
+        current.next = Node(value)
+        
 
     # method to return the max value in the linked list
     # returns the data value and not the node
     def find_max(self):
```
Comment (spitsfire) [inline linked_list/linked_list.py:114]: 👍

File: linked_list/linked_list.py [hunk 14]
```diff
@@ -13,63 +13,140 @@ def __init__(self):
...
+        while current:
+            if current.value > max:
+                max = current.value 
+            current = current.next
+        
+        return max
 
     # method to delete the first node found with specified value
     # Time Complexity: ?
     # Space Complexity: ?
```
Comment (spitsfire) [inline linked_list/linked_list.py:130]: What do you think the space and time complexity of this is? Is the time complexity dependent on the length of the linked list? Is there any new variables being created that grow larger depending upon the input?

File: linked_list/linked_list.py [hunk 15]
```diff
@@ -13,63 +13,140 @@ def __init__(self):
...
     def delete(self, value):
-        pass
+        current = self.head
+ 
+        if current is not None:
+            if current.value == value:
+                self.head = current.next
+                temp = None
+                return
+
```
Comment (spitsfire) [inline linked_list/linked_list.py:139]: We can shorten this up a bit to make it more readable:

```suggestion
        if self.head.value == value:
            self.head = current.next
            return
                
         current = self.head
```

File: linked_list/linked_list.py [hunk 16]
```diff
@@ -13,63 +13,140 @@ def __init__(self):
...
+                break
+            prev = current
+            current = current.next
+ 
+        if current == None:
+            return None
+ 
+        prev.next = current.next
+ 
+        current = None
```
Comment (spitsfire) [inline linked_list/linked_list.py:151]: We can shorten this, too, to make it more readable:

```suggestion
        while current is not None:
            if current.next.value == value:
                current.next = current.next.next
            current = current.next
```

We don't need to re-assign current to `None` nor return `None`

File: linked_list/linked_list.py [hunk 17]
```diff
@@ -86,32 +163,84 @@ def visit(self):
 
     # method to reverse the singly linked list
     # note: the nodes should be moved and not just the values in the nodes
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(n)?
+    # Space Complexity: O(1)?
     def reverse(self):
```
Comment (spitsfire) [inline linked_list/linked_list.py:170]: 👍

File: linked_list/linked_list.py [hunk 18]
```diff
@@ -86,32 +163,84 @@ def visit(self):
...
     def find_middle_value(self):
-        pass
+        if self.head == None:
+            return None
+
+        current = self.head
+        count = 0
+        while current:
+            count += 1
+            current = current.next
```
Comment (spitsfire) [inline linked_list/linked_list.py:193]: hmm where have we seen this already? Maybe we can call `length` to get the number of nodes

File: linked_list/linked_list.py [hunk 19]
```diff
@@ -86,32 +163,84 @@ def visit(self):
...
+
+        
+        
 
     # checks if the linked list has a cycle. A cycle exists if any node in the
     # linked list links to a node already visited.
     # returns true if a cycle is found, false otherwise.
     # Time Complexity: ?
     # Space Complexity: ?
     def has_cycle(self):
```
Comment (spitsfire) [inline linked_list/linked_list.py:232]: 👍 interesting approach!

File: linked_list/linked_list.py [hunk 20]
```diff
@@ -86,32 +163,84 @@ def visit(self):
...
+        return current.value
+
+        
+        
 
     # checks if the linked list has a cycle. A cycle exists if any node in the
     # linked list links to a node already visited.
     # returns true if a cycle is found, false otherwise.
     # Time Complexity: ?
     # Space Complexity: ?
```
Comment (spitsfire) [inline linked_list/linked_list.py]: What do you think the space and time complexity of this is? Is the time complexity dependent on the length of the linked list? Is there any new variables being created that grow larger depending upon the input?

**Review by spitsfire — COMMENTED** [review]
You are missing a lot of your space/time complexity explanations

File: linked_list/linked_list.py (hunk 1 above)
Comment (waterlilly169) [reply linked_list/linked_list.py:18]: resubmitted with questions filled out

**Comment by waterlilly169** [comment]
Resubmitted with time complexity
